system/jlib/jptree.cpp

```diff
@@ -9481,3 +9542,341 @@ void copyPropIfMissing(IPropertyTree & target, const char * targetName, IPropert
+{
+    if (hexsequence)
+    {
+        //close current hex stream sequence
+        s.append('_');
+        hexsequence = false;
+    }
+}
+inline bool isValidPTreeNameChar(const char *ch, unsigned remaining, unsigned &chlen, bool isatt, bool first)
+{
+    chlen = 0;
+    if (!ch || !*ch || !remaining)
+        return false;
+    chlen = utf8CharLen((const unsigned char *)ch, remaining);
+    if (0==chlen)
+    {
+        chlen=1;
+        return false;
+    }
+    if (chlen>1)
+        return !isatt;  //attributes do not seem to currently support multibyte characters in their names
+    if (first)
+        return isValidXPathStartChr(*ch);
+    return isValidXPathChr(*ch);
+}
+
+const char *findFirstInvalidPTreeNameChar(const char *ch, unsigned remaining)
+{
+    if (!ch)
+        return nullptr;
+    if (remaining==0 || *ch=='\0') //empty name still needs to trigger a special encoding
+        return ch;
+    bool isatt = ('@'==*ch);
+    if (isatt)
+    {
+        ch++;
+        remaining--;
+    }
+    if (!remaining)
+        return nullptr;
+
+    unsigned chlen = 0;
+    if (!isValidPTreeNameChar(ch, remaining, chlen, isatt, true))
+        return ch;
+    if (remaining<chlen)
+        return nullptr;
+    ch += chlen;
+    remaining -= chlen;
+    while (remaining)
+    {
+        if (!isValidPTreeNameChar(ch, remaining, chlen, isatt, false))
+            return ch;
+        if (remaining<=chlen)
+            return nullptr;
+        ch += chlen;
+        remaining -= chlen;
+    }
+    return nullptr;
+}
+
+const char *findFirstInvalidPTreeNameChar(const char *ch)
+{
+    if (!ch)
+        return nullptr;
+    return findFirstInvalidPTreeNameChar(ch, strlen(ch));
+}
+
+static inline bool checkAppendValidPTreeNameChar(StringBuffer &s, const char *ch, unsigned remaining, unsigned &chlen, bool isatt, bool first, bool &hexsequence)
+{
+    if (!isValidPTreeNameChar(ch, remaining, chlen, isatt, first))
+        return false;
+    checkEndHexSequence(s, hexsequence);
+    s.append(chlen, ch);
+    return true;
+}
+
+inline void appendPTreeNameSpecialEncoding(StringBuffer &s, const char *enc, bool &hexsequence)
+{
+    checkEndHexSequence(s, hexsequence);
+    s.append('_');
+    if (!isEmptyString(enc))
+        s.append(enc);
+}
+
+inline StringBuffer &encodePTreeNameUtf8Char(StringBuffer &s, const char *&ch, unsigned &remaining, bool isatt, bool &hexsequence, bool &first)
+{
+    unsigned chlen = 1;
+    if (first && (*ch=='@'))
+    {
+        s.append('@');
+        ch++;
+        remaining--;
+        if (remaining==0 || *ch=='\0')
+            return s;
+    }
+    if (*ch!='_' && checkAppendValidPTreeNameChar(s, ch, remaining, chlen, isatt, first, hexsequence))
+    {
+        remaining -= chlen;
+        ch += chlen;
+        first = false;
+        return s;
+    }
+    switch (*ch)
+    {
+        case '_':
+            appendPTreeNameSpecialEncoding(s, "_", hexsequence);
```

Comment:
may be worth a variety of appendPTreeNameSpecialEncoding that doesn't check isEmptyString(enc) on each call to use here, since it is exlpicitly not empty.

Reply:
appendPTreeNameSpecialEncoding no longer checks isEmptyString().